Next generation Tile Reduce

mourner opened the pull request:
An early version of the next-generation Tile Reduce that is **thousands of times faster** than the current one. This is _not_ a joke. 

For example, generating GeoJSON road difference between OSM & Tiger 2015 for US West z12 (~26% of US) takes **2 minutes 45 seconds** on my Macbook Pro; previously it would take **a day** on a much more powerful EC2 c3.4xlarge instance eating all of the available RAM. 

The new architecture also makes Tile Reduce much more flexible and loosely coupled, e.g. easy to extend for raster support, weird tile formats, lazy decoding, etc. 

I'll devlog how I did this, but the main idea behind the rewrite is removing almost all communication between the main thread and the workers and avoiding unnecessary work:

- responsibility for fetching/parsing tiles fully moves to the workers
- vector tiles are not parsed and converted to GeoJSON by default; instead, VT layer objects are passed to the workers for lazy decoding (parsing only the features that are relevant to the current task)
- big GeoJSON output like difference is never passed back to the main thread; instead, it's streamed directly from workers to stdout; `reduce` event is only for smaller values like statistics

Somewhat breaks API compatibility, but this is something we have to do to support the new architecture. Upgrading existing processors should be easy once we finalize the work.

Also sports a fancy progress bar so that you could _feel_ the unreal performance:

![tilereduce-progress](https://cloud.githubusercontent.com/assets/25395/10845973/b7cfc43e-7f14-11e5-9e6c-84d42d997cab.gif)

While all the basic stuff works, this is still a work in progress. Rough list of remaining tasks:

- [x] clean up and push the new diff processor example (today)
- [x] get feedback on the API changes
- [x] code review by the team
- [x] readd remote tiles support
- [x] readd non-bbox covers
- [x] streaming tile list input
- [x] avoid overloading the message queue
- [x] ~~readd overzoom (slicing to higher zoom levels than the sources provide)~~
- [x] fix interleaved output issue @tcql
- [x] readd tests (in progress) @morganherlocker 
- [x] readd rate limiting
- [x] support streaming tile list from mbtiles directly @mourner
- [x] documentation & upgrade guide

I would love feedback and help on this. We could win so much with a blazing fast Tile Reduce, so let's make it happen sooner.

cc @morganherlocker @MateoV @aaronlidman @tcql @tmcw @lxbarth @ericfischer @batpad @Rub21 @geohacker @arunasank @isiyu

mourner commented:
[Here's that US West OSM-Tiger diff](https://api.mapbox.com/styles/v1/mourner/cigdrt1in0000b6lyla5codqz.html?title=true&access_token=pk.eyJ1IjoibW91cm5lciIsImEiOiJWWnRiWG1VIn0.j6eccFHpE3Q04XPLI7JxbA#5.12/38.789/-114.270) I generated in 3 minutes on a map (disregard the weird boundaries).

aaronlidman commented:
This sounds fantastic Vlad, I'll put aside some time this weekend to look more in-depth.

tcql commented:
just adding my :+1: / proof; this is way fast. I did tests using mapbox/osm-sidewalker. I like it. The api on the `map` side feels slightly less friendly to newbies (since it comes through as vector tile instead of geojson), but I think it's advantageous overall. 

I'm definitely in to sink some time to help move this too.

mourner commented:
> The api on the map side feels slightly less friendly to newbies (since it comes through as vector tile instead of geojson), but I think it's advantageous overall.

With the new architecture, it's pretty easy to make an option to pass GeoJSON instead of VT to the map function, and make it the default. It will still be relatively fast.

tcql commented:
added support for tile queue pausing, so we don't pass too many messages to child workers

mourner commented:
@tcql manual limiting broke the tile-reduce — now if you run the count example, the total tiles and results are different each time.

mourner commented:
Fixed the tile list text stream somehow, but the JS array tile stream is still broken — for some reason it finishes when it hits the 50k limit.

mourner commented:
Fixed the JS stream as well. Now I wonder if there would be a benefit to try to switch to using `spawn` and piping between the processes instead of `send`, like in comments here https://github.com/nodejs/node-v0.x-archive/issues/4374

mourner commented:
Running count on z12 world on my MBP, it goes super-smooth and fast for a while (getting to maybe 30% of the world tiles), then suddenly node workers memory starts to grow out of control and it starts to stutter, eventually filling the memory and making the terminal unresponsive:

![image](https://cloud.githubusercontent.com/assets/25395/10970249/1a03374c-83d6-11e5-9a6f-6b4601f30b6b.png)

Middle flat green area is the job at first, then you see the spike.

I really wonder what might be the cause. We want the runs to have flat memory consumption regardless of the job size.

mourner commented:
Ha, when I switched the pause limit from 50000 to 5000, the world count went smoothly without memory spikes and finished successfuly in under 7 minutes!

```
2303777 / 2303777 tiles (100%), 401.8s elapsed [===============...
Features total: 436992923
```

Now let's try OSM coverage...

mourner commented:
> if we handled process communication with streams, backpressure would stop the problem, but we'd kind of need to do a file readstream => some round-robin selection of which worker to pipe to

Now I see what you mean here @tcql (I'm still a streams noob), let's explore this cool idea. Another benefit is that as far as I understand, in pre-Node4 forked process send is blocking while spawned process piping is not. And we could get some perf gains by fully avoiding spamming the event loop.

mourner commented:
@tcql nope, replacing `isPaused()` with `paused` still breaks everything in Node 0.12.

morganherlocker commented:
I am working on testing now. I have generated a very small mbtiles fixture weighing in at 1MB total. I think it will simplify our process if we just check this in for testing, rather than dealing with s3 and caching fixture assets in our CI. For benchmarking (more rare), people can use whatever they deem appropriate, such as the world OSM QA mbtiles file, which is just a `curl` away.

https://api.mapbox.com/styles/v1/morganherlocker/cigshqusv005xa6ks7iua9b2l.html?title=true&access_token=pk.eyJ1IjoibW9yZ2FuaGVybG9ja2VyIiwiYSI6Ii1zLU4xOWMifQ.FubD68OEerk74AYCLduMZQ#15.4/36.9777/-122.0222/-0.3

<img width="1145" alt="screenshot 2015-11-09 14 52 28" src="https://cloud.githubusercontent.com/assets/814300/11049189/8a37a7f0-86f1-11e5-9dac-3746d7ccbb79.png">

morganherlocker commented:
I processed an old Tiger road file from 2009 for the same area so we can test data with multiple sources, and have the diffing example be runnable without downloading and processing external data. These are published at a stable mapid, so we can use that url for testing http tile sources.

<img width="1184" alt="screenshot 2015-11-09 16 00 52" src="https://cloud.githubusercontent.com/assets/814300/11050495/57e21632-86fb-11e5-9ab6-d9a6991e0afc.png">

morganherlocker commented on src/worker.js:
```diff
@@ -0,0 +1,49 @@
+});
+
+function loadSource(source, done) {
+  if (source.mbtiles) sources.push({name: source.name, getTile: require('./mbtiles')(source, done)});
```
Are we waiting for the `ready` callback to be called anywhere? For the `remote` parser it appears that the getTile parser is not actually returned by the time the `ready` callback fires. I am not sure if this is a bug, but the flow is a bit difficult to follow.

mourner replied:
@morganherlocker for the remote parser, ready is called immediately because there is no initialization step — it is ready to receive requests immediately unlike mbtiles which needs to open a db connection first before getting any tiles to work on.

morganherlocker replied:
I wonder if there is a way to make this an interface where any source can be used the same way, without knowing the implementation details (this could help us with additional plugins down the road). Perhaps instead of returning the getTile function synchronously, we should send it back as the result of the `ready` callback. I believe that would allow users of the parse interfaces to treat them the same.

morganherlocker replied:
Ok, I am going to proceed with writing tests for this in its current form and may refactor in a separate PR (requires some refactoring of the worker implementation). The remote interface  diverts the execution flow in a confusing way IMO, for example:

```js
var getTile = remote(source, function(err) {
  // i need to wait until i get here to see if there is an error on initialization
  if(err) throw err;
  // next line will fail because getTile has not actually returned yet, despite the parser being "ready"
  getTile([5276, 12757, 15], function(err, layers) {
    // yay, results!
  });
});
```

mourner replied:
I agree, we should improve the flow here to make things easier to test and extend.

mourner replied:
Implemented your suggestion.

morganherlocker commented:
@mourner it appears we return all layers in this branch, as opposed to only the requested layers in the source object. Could preserving the `layers` array provide any significant parsing gains (keep in mind most people will want geojson most of the time)? Is selectively parsing worth the effort on the VT parsing side?

Current source example:

```
{
  name: 'streets',
  url: 'https://b.tiles.mapbox.com/v4/mapbox.mapbox-streets-v6/{z}/{x}/{y}.vector.pbf',
  layers: ['roads', 'tunnels']
}
```

morganherlocker commented:
The latest build is passing on node 0.10, but failing on node 0.12. I added 0.10 to the travis config to track this behavior. 

The issue is that the decoded vector tile under 0.10 looks like this with a raw data array:

```
{"buildings":{"version":1,"name":"buildings","extent":2048,"length":264,"_pbf":{"buf":[26,243,128,1,10,9,98,117, [...]
```

..while the 0.12 version returns a Buffer object with a wrapped data array:

```
{"buildings":{"version":1,"name":"buildings","extent":2048,"length":264,"_pbf":{"buf":{"type":"Buffer","data":[26,243,128,1,10,9,98,117,105,108,100,105,110,103,115,18,40,18,6,0,0,1,1,2,2,24,3 [...]
```

My hunch is that this issue is upstream in either the `pbf` lib or the `vector-tile-js` lib, but I will need to do some more digging.

mourner commented:
@morganherlocker regarding `layers` — I didn't readd it yet because of lazy pbf decoding, but when we readd automatic GeoJSON parsing by default, we should definitely also readd `layers` property at the same time for performance gains (but make it optional so that it parses all layers by default).

Regarding VT 0.10 vs 0.12, I'm looking into the issue now.

mourner commented:
@morganherlocker figured out the VT decoding issue — this happens simply because JSON.stringify has a different behavior on Buffer objects:

```js
console.log(JSON.stringify(new Buffer([0xc8,0xe8,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0x01])));
```

```bash
[200,232,255,255,255,255,255,255,255,1] # node 0.10
{"type":"Buffer","data":[200,232,255,255,255,255,255,255,255,1]} # node 0.12
```

You just need to compare objects differently, e.g. using `t.deepEqual` on objects rather than comparing strings.

mourner commented:
Added an ability to omit area alltogether if you use mbtiles sources — it'll stream tile numbers from the first source if it's mbtiles using node-mbtiles ZXYStream. We can update the API later to be able to choose which source to use.

morganherlocker commented:
I am noticing that binarysplit seems to crash when tile-reduce is run in the context of a test. I am pretty sure the main thread is grabbing TAP data somewhere, but need to do some more digging.

morganherlocker commented:
Passes:

```
tape test/test.*.js
```

Fails: 

```
tap test/test.*.js
```
Error:

`TypeError: Object #<Socket> has no method 'cursorTo'`

It looks like somehow tap is injecting some data that is throwing off the progress display.

morganherlocker commented:
>  readd rate limiting

I am wondering if this will be easiest to implement as a *per thread* rate limit. Coordinating a global rate limit across works will be a pain and likely unreliable (ran into this the first time around and irrc it was the reason I stuck with single threaded source parsing). We could take one master rate limit parameter and simply allot an equal fraction of the budget to each worker.

morganherlocker commented:
The current tile logic automatically uses the mbtiles stream of the first source. To get to geojson, bbox, or tile array, you have to set `options.tiles` to `true`.

```js
if (!options.tiles && options.sources[0].mbtiles) {
  // mbtiles zxystream
  var db = new MBTiles(options.sources[0].mbtiles, function(err) {
    if (err) throw err;
    tileStream = db.createZXYStream({batch: pauseLimit}).pipe(binarysplit()).on('data', handleZXYLine);
  });
} else if (typeof options.tiles === 'string') {
  // text file tile stream ("x y z\n")
  tileStream = fs.createReadStream(options.tiles);
  tileStream.pipe(binarysplit()).on('data', handleTileLine);
} else {
  // JS tile array, GeoJSON or bbox
  tileStream = streamArray(cover(options.tiles)).on('data', handleTile);
}
```

How I think it should work is that there are set of accepted options:

- [ ] `options.tiles`: tile array (string or array, but I think we should explicitly just choose array)
- [ ] `options.bbox`: a bbox array
- [ ] `options.geojson`: a geojson object that will be tile-cover'ed
- [ ] `options.tileStream`: a tile stream
- [ ] `options.mbtilesCover`: path to an mbtiles file that contains the geometry you want covered (don't default to the first one; I think people will find this difficult to debug without reading the API docs closely)

We could use the first valid option encountered, or we could throw an error if more than one are encountered. I am leaning towards throwing an error, since it removes any possible ambiguity.

mourner commented:
> The current tile logic automatically uses the mbtiles stream of the first source. To get to geojson, bbox, or tile array, you have to set options.tiles to true.

Ah you're right, missed that. Need to work on more test coverage of input options.

>  options.tiles: tile array (string or array, but I think we should explicitly just choose array)

If we explicitly choose JS array, do we pass the string (text file path) as a different option? I thought of something like `tilesTextList` or something but looked a bit ugly, so I settled with string `tiles`.

> options.tileStream: a tile stream

If this means a Node readable stream of tiles, — I tried doing this but it turned out to be problematic, since a stream will often start emitting data before we catch it, or before we're ready to consume tile numbers. That's a Node streams quirk me and @tcql still find very confusing.

> options.mbtilesCover: path to an mbtiles file that contains the geometry you want covered

I think we should be consistent here — either `geojsonCover` and `mbtilesCover`, or `geojson` and `mbtiles`. Or maybe `coverGeoJSON` and `coverMBTiles`.

> (don't default to the first one; I think people will find this difficult to debug without reading the API docs closely)

I would prefer leaving the default in some form — it makes the API easier, because most of the time you would just want "all the tiles in the input source" anyway, and specifying the same path in two places looks a bit verbose and redundant. When you specify multiple sources, you usually want "tiles that are present in both", and that's what it does currently — streaming tiles from one source and checking whether they exist in others, and only running the map function on those.

If we want to be more explicit, how about this logic:

- if we only provide a _single_ mbtiles source and don't provide area/tiles in some form, it automatically uses all available tiles from that source
- if we provide multiple sources without providing the area, you can specify something like `mbtilesCover: 'osm'` in the options, referring to the tile source by name explicitly, otherwise it'll either use the first encountered mbtiles source (my preference), or just throw and error asking to specify.

> We could use the first valid option encountered, or we could throw an error if more than one are encountered. I am leaning towards throwing an error, since it removes any possible ambiguity.

It may be nice to throw an error on more than one, but it may be cumbersome to implement in code, so I'd settle for "look for options in a defined order" for now and maybe improve later. One good way to help with ambiguity would be to just output a more detailed log in the console before starting the progress count, e.g.:

```bash
Starting tile-reduce on bbox: [40, 40, 50, 50]...
Starting up 8 workers... 
Computing cover...
60143 tiles processed in 9s.
```

tcql commented:
> If this means a Node readable stream of tiles, — I tried doing this but it turned out to be problematic, since a stream will often start emitting data before we catch it, or before we're ready to consume tile numbers.

I still want to take another stab at getting this to behave. I'll try to play around today and see if I can prototype something.

> If we want to be more explicit, how about this logic:
>
> - if we only provide a single mbtiles source and don't provide area/tiles in some form, it automatically uses all available tiles from that source
> - if we provide multiple sources without providing the area, you can specify something like mbtilesCover: 'osm' in the options, referring to the tile source by name explicitly, otherwise it'll either use the first encountered mbtiles source (my preference), or just throw and error asking to specify.

:+1: I think I like referencing the source by name. I can't think of a good reason why you'd want to get a tilelist out of an mbtiles that aren't also included as a source. I think in an ideal world this would be not specifically `mbtilesCover`, but `sourceCover`, but that depends highly on what kind of sources we implement, and if they support fetching tilelists.

> We could use the first valid option encountered, or we could throw an error if more than one are encountered. I am leaning towards throwing an error, since it removes any possible ambiguity.

:+1:

mourner commented:
Updated the logic to the one described above and added `sourceCover` option to be able to explicitly specify the source to use. Now I need to move all stream generation code to cover.js, and cover that with tests.

mourner commented:
@morganherlocker also removed the VT parsing duplication.

mourner commented:
> We could take one master rate limit parameter and simply allot an equal fraction of the budget to each worker.

@morganherlocker yes, let's do it the simple way. Initially I thought we could just remove rate limiting by setting `maxSockets` option in http which limits the number of concurrently running connections and seems like a more fair limit, but I'm not sure now.

mourner commented:
This is starting to look really solid. The remaining tasks I see before we release next week:

- [ ] bumping test coverage from the current 69% to around 100% (branch)
- [ ] cleaning up documentation
- [ ] (optional) explicit remote source rate limiting (not as critical since it's also limited by maxSockets)

We can handle minor code/API inconsistencies and features with subsequent minor releases.

tcql commented:
just for visibility; as of f1a590b, we dropped support for `tiles` taking a string for a file, instead we now support `tileStream`, which is any stream of `x y z` strings or `[x, y, z]` arrays. Users can easily read from a tile file, stdin, or any other streaming resource

mourner commented:
Coverage is up to 95% statements, 86% branches, which is good enough for a release.

morganherlocker commented:
We are looking really good here on features, tests, speed, and docs. I want to :ship: tomorrow afternoon with a 3.0.0 tag.

- [ ] @tcql - :eyeglasses: on doc edits I made?
- [ ] @mourner - :eyes: final pass code review for 3.0?
- [ ] @anandthakker - voice of reason; any glaring overlooks you can spot here? (take a look at the changelog on this PR for an overview)
- [ ] @morganherlocker - write a couple practical examples (and rm -rf the current diff)

mourner commented:
Reviewed, all looks good! One minor thing is that remote throttle test runs for too long (I'd give it not more than a few seconds), but we can deal with that later. Looks like the PR also needs rebasing against master now — I can take care of that.

mourner commented:
Warning: force-pushed a rebased branch. It merges cleanly now. Also squashed some tiny commits like "fix typo".

mourner commented:
The diff script now lives at https://github.com/mourner/road-diff. I'll clean it up before blogging next week.

tcql commented:
working on adding a short doc section about events, so we've got bases covered

mourner commented:
:+1: let's merge and publish 3.0.0. We can improve as we go after that.

morganherlocker commented:
:shipit: 

![](http://media3.giphy.com/media/dkGhBWE3SyzXW/giphy.gif)

lxbarth commented:
:steam_locomotive: !!
